src/Middleware/WebSockets/src/HandshakeHelpers.cs

```diff
@@ -110,23 +110,20 @@ public static string CreateResponseKey(string requestKey)
             // this concatenated value to obtain a 20-byte value and base64-encoding"
             // https://tools.ietf.org/html/rfc6455#section-4.2.2
 
-            using (var algorithm = SHA1.Create())
-            {
-                // requestKey is already verified to be small (24 bytes) by 'IsRequestKeyValid()' and everything is 1:1 mapping to UTF8 bytes
-                // so this can be hardcoded to 60 bytes for the requestKey + static websocket string
-                Span<byte> mergedBytes = stackalloc byte[60];
-                Encoding.UTF8.GetBytes(requestKey, mergedBytes);
-                EncodedWebSocketKey.CopyTo(mergedBytes.Slice(24));
-
-                Span<byte> hashedBytes = stackalloc byte[20];
-                var success = algorithm.TryComputeHash(mergedBytes, hashedBytes, out var written);
-                if (!success || written != 20)
-                {
-                    throw new InvalidOperationException("Could not compute the hash for the 'Sec-WebSocket-Accept' header.");
-                }
+            // requestKey is already verified to be small (24 bytes) by 'IsRequestKeyValid()' and everything is 1:1 mapping to UTF8 bytes
+            // so this can be hardcoded to 60 bytes for the requestKey + static websocket string
+            Span<byte> mergedBytes = stackalloc byte[60];
+            Encoding.UTF8.GetBytes(requestKey, mergedBytes);
+            EncodedWebSocketKey.CopyTo(mergedBytes.Slice(24));
 
-                return Convert.ToBase64String(hashedBytes);
+            Span<byte> hashedBytes = stackalloc byte[20];
+            var written = SHA1.HashData(mergedBytes, hashedBytes);
```

Comment:
Does `SHA1.HashData` allocate any objects or is it truly static?

Reply:
I believe it only allocates on Win7 & Win8. But it's still faster than alternatives: see https://github.com/dotnet/runtime/pull/40510.

Reply:
I care more about linux? Is it allocating? This change had a measurable impact on websocket upgrades when it was made.  I want to make sure this isn't regressing that.

Reply:
@BrennanConroy did you commit those performance tests?

Reply:
Of course 😄 
https://github.com/dotnet/aspnetcore/blob/master/src/Middleware/perf/Microbenchmarks/HandshakeBenchmark.cs

Reply:
Levi can you run these and show the before after?

Reply:
Last I checked Linux was neither Win7 nor Win8, so it should be non-allocating. :)

Here's the Linux implementation, for reference:
https://github.com/dotnet/runtime/blob/46acbaa5f13b501dc2e750bc2023d4a30eded9c7/src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Unix.cs#L43-L62

Reply:
I don't have a suitable environment for running the perf tests, sorry. If you have a shared environment where I can do so, along with clear instructions for it (copy + paste these exact commands), I can give it a spin.

Reply:
Unrelated to this PR, if _IsRequestKeyValid_ is a hot method I think you can get significant perf gains in it from rolling your own logic. The framework method you're calling performs base64 decoding, but you don't actually need to decode it. You only care that it is valid base64 and that it decodes to 16 bytes. So a rough outline of the logic would be:

1) Validate that the incoming string length is _exactly_ 24 chars.
2) Perform vectorized logic to validate that the first 22 chars of the string are within `[a-zA-Z0-9+/]`. (Your vectorized logic will consume 8 or 16 chars at a time, depending on the current OS.)
3) Validate that the last 2 chars are both `'='`.

If any of these checks fail, you can fall back to the framework's `Convert.TryFromBsae64String` string for extra validation if you need to handle things like whitespace chars.

Reply:
Linux

Before:
|            Method |       Mean |    Error |   StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |-----------:|---------:|---------:|------------:|-------:|------:|------:|----------:|
| CreateResponseKey | 2,292.6 ns | 22.81 ns | 21.33 ns |   436,191.8 | 0.0031 |     - |     - |     200 B |
| IsRequestKeyValid |   143.3 ns |  1.13 ns |  1.00 ns | 6,979,003.0 |      - |     - |     - |         - |

After:
|            Method |       Mean |    Error |   StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |-----------:|---------:|---------:|------------:|-------:|------:|------:|----------:|
| CreateResponseKey | 1,406.8 ns | 12.49 ns | 11.07 ns |   710,842.7 | 0.0015 |     - |     - |      80 B |
| IsRequestKeyValid |   147.2 ns |  1.12 ns |  0.99 ns | 6,792,335.8 |      - |     - |     - |         - |

Reply:
> if IsRequestKeyValid is a hot method I think you can get significant perf gains in it from rolling your own logic

We had a try of this in https://github.com/dotnet/aspnetcore/pull/13196